functions/report-psychometrician-create-download/service/psychometrician-report.service.js

```diff
@@ -0,0 +1,158 @@
+  },
+
+  /**
+   * Return the CSV file as a string
+   * @return {Promise<void>}
+   */
+  generatePsychometricianReport: async function generatePsychometricianReport () {
+    // Read data from the cache
+    const results = await psychometricianReportDataService.sqlFindAllPsychometricianReports()
+    const output = []
+    for (const obj of results) {
+      output.push(obj.jsonData)
+    }
+    const headers = this.producePsychometricianReportDataHeaders(results)
+
+    return new Promise((resolve, reject) => {
+      csv.writeToString(
+        output,
+        { headers: headers },
+        function (err, data) {
+          if (err) { return reject(err) }
+          resolve(data)
+        }
+      )
+    })
+  },
+
+  /**
+   * Return the CSV file as a string
+   * @return {Promise<void>}
+   */
+  generateAnomalyReport: async function generateAnomalyReport () {
+    // Read data from the cache
+    const results = await psychometricianReportDataService.sqlFindAllAnomalyReports()
+    const output = []
+    for (const obj of results) {
+      output.push(obj.jsonData)
+    }
+
+    const headers = this.produceAnomalyReportDataHeaders()
+    output.unshift(headers)
+
+    return new Promise((resolve, reject) => {
+      csv.writeToString(
+        output,
+        { headers: false },
+        function (err, data) {
+          if (err) { return reject(err) }
+          resolve(data)
+        }
+      )
+    })
+  },
+
+  /**
+   * Generate and upload the psychometrician and anomaly reports in Zip format to Azure Storage
+   * This is a lift and shift version: it needs to be updated
+   * TODO: handle volume data
+   * @return {Promise<void>}
+   */
+  process: async function process () {
+    const dateGenerated = moment()
+    const psychometricianReport = await this.generatePsychometricianReport()
+    const anomalyReport = await this.generateAnomalyReport()
+    const zipFile = await this.generateZip(psychometricianReport, anomalyReport, dateGenerated)
+    const uploadBlob = await this.uploadToBlobStorage(zipFile)
+    const md5Buffer = Buffer.from(uploadBlob.contentSettings.contentMD5, 'base64')
+    await psychometricianReportDataService.sqlSaveFileUploadMeta(
+      uploadBlob.container,
+      uploadBlob.name,
+      uploadBlob.etag,
+      md5Buffer,
+      psychometricianReportCode)
+  },
+
+  /**
+   * Returns the CSV headers
+   * @param {Array} results
+   * @returns {Array}
+   */
+  producePsychometricianReportDataHeaders: function producePsychometricianReportDataHeaders (results) {
+    // If there are no checks, there will be an empty file
+    if (results.length === 0) return []
+    // Fetch the first completed check to store the keys as headers
+    const completedCheck = results.find(c => c.jsonData && c.jsonData.hasOwnProperty('Q1ID'))
+    if (completedCheck) {
+      return Object.keys(completedCheck.jsonData)
+    }
+    // Alternatively return the first check keys
```

Comment:
what is the impact of returning incomplete headers in that case?
will that cause any data inconsistency?

Reply:
See my previous comment about this being a lift and shift.  It appears to be defensive programming.